vault/logical_raw.go

```diff
@@ -110,11 +138,76 @@ func (b *RawBackend) handleRawWrite(ctx context.Context, req *logical.Request, d
 		}
 	}
 
-	value := data.Get("value").(string)
+	v := data.Get("value").(string)
+	value := []byte(v)
+	if encoding == "base64" {
+		var err error
+		value, err = base64.StdEncoding.DecodeString(v)
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+		}
+	}
+
+	if compressed {
```

Comment:
If the user doesn't provide compressed=true, we're not making use of the information we have in the existing storage entry, and we'll go ahead and replace an existing compressed entry with an uncompressed one.  Can you reorder things so that if this is an update (instead of a create), we use the information in the existing storage entry to determine whether and how to compress?  And in that case maybe we should return an error if they did specify those options and they don't match the existing storage entry?

Reply:
Just clarifying.   To differentiate between update vs create, backends generally need to implement "ExistenceCheck" -- https://github.com/hashicorp/vault/blob/main/sdk/framework/path.go#L92.

Are you saying you want this diff to implement "ExistenceCheck" for sys/raw and use that in "handleRawWrite", or skip ExistenceCheck and have "handleRawWrite" do something like

			
entry, err := b.barrier.Get(ctx, path)
_, compressionType, _, _ := compressutil.DecompressWithCanary(entry.Value)
if compressionType != "" && !compressed {
    return err // this is compressed data but compression was not specified
}

One problem with this approach is what if users **want** to replace an existing compressed entry with an uncompressed one (that's actually what happens today if you use sys/raw to update any json-compressed entries like core/audit or core/mounts).

Reply:
Okay, I implemented an existence check -- and updates now will use the existing compression type (or no compression) -- and check + error out if the new compression type is different the existing compression type.

I did make one small exception, users can force an entry to be non-compressed even if it was compressed before.  This ensures a users can workaround unknown issues with compressutil .